ENH Updated EY Notebook

LeJit opened the pull request:
This is an in-progress version of the [EY whitepaper](https://www.microsoft.com/en-us/research/uploads/prod/2020/09/Fairlearn-EY_WhitePaper-2020-09-22.pdf) notebook. The main goal of this notebook is to highlight some of the functionality of Fairlearn (e.g. how to select an inner model of ExponentiatedGradient, how to incorporate uncertainty quantification).

The goal of this draft pull request is to community feedback about the current workflow and whether this use case aligns with our community values. Some concerns that have been surfaced:
- We introduce a synthetic feature `LIMIT` to generate a disparity between `Male` and `Female` applicants in the model's performance.
- We are using a dataset pertaining to loans in Taiwan (collected in 2005) to motivate an example around credit lending in the United States.

Items still left to complete:
- Add conclusion sections.
- Add problem introduction and motivation section.

romanlutz commented:
Are you sure you want to put it in an ipynb, not .py in the examples folder?

LeJit commented:
Miro and I discussed leaving it in .ipynb format for now and converting to a .py file at a later time. In my opinion, the SciPy tutorial/PyData Global notebook is a better fit for the examples folder.

romanlutz commented:
> Miro and I discussed leaving it in .ipynb format for now and converting to a .py file at a later time. In my opinion, the SciPy tutorial/PyData Global notebook is a better fit for the examples folder.

Thanks for elaborating! I agree on converting other use cases to `.py`, but that doesn't necessarily mean we should leave this one in `.ipynb`. That said, perhaps this is a separate issue that can be tackled in another PR. I was probably optimistically thinking we'd do all of those tasks at one go. I'm not sure whether we should add more notebooks into that folder, though. Not converting to `.py` is one thing, but adding even more that eventually need converting is not a step in the right direction IMO. I'm all in favor of fixing/updating existing ones, of course.

Tagging @MiroDudik since you mentioned that there was some prior discussion so that he's aware.

MiroDudik commented:
> > Miro and I discussed leaving it in .ipynb format for now and converting to a .py file at a later time. In my opinion, the SciPy tutorial/PyData Global notebook is a better fit for the examples folder.
> 
> Thanks for elaborating! I agree on converting other use cases to `.py`, but that doesn't necessarily mean we should leave this one in `.ipynb`. That said, perhaps this is a separate issue that can be tackled in another PR. I was probably optimistically thinking we'd do all of those tasks at one go. I'm not sure whether we should add more notebooks into that folder, though. Not converting to `.py` is one thing, but adding even more that eventually need converting is not a step in the right direction IMO. I'm all in favor of fixing/updating existing ones, of course.
> 
> Tagging @MiroDudik since you mentioned that there was some prior discussion so that he's aware.

I think that the idea is that the notebook in this PR would replace this one: https://github.com/fairlearn/fairlearn/blob/main/notebooks/Binary%20Classification%20with%20the%20UCI%20Credit-card%20Default%20Dataset.ipynb

I agree that we should discourage creating new notebooks in `.ipynb`, but I'm viewing this specific PR as a replacement of an existing notebook, so I'm okay with it being `.ipynb`. This is being linked to from our landing page, so I'm in favor of expediency here.

romanlutz commented:
That works for me.  I see one newly added ipynb and one changed one, though. Are we planning to delete one of these? @LeJit 

Perhaps tangential, but I'm really a fan of the notebooks you (and others) created for SciPy and PyData Global. If we ever want to link to those I'd be very much in favor.

LeJit commented:
Yes. The end goal is to delete the old notebook, eventually. I left it in for now, so that people have something to compare the new notebook to.

LeJit commented:
@MiroDudik The final "draft" is ready.

mmadaio reviewed (commented):
Thanks for this, Manojit! 

I think at a high-level, this will be a helpful use case for users! My main comment is that there's a lot here around what you did, but less explanation provided as to why certain decisions were made. 

**It would be good to add more explanation as to why you're using "sex" as the sensitive feature**. 
Is this because of historical inequities in lending to women? Or prior evidence of disparities in loan prediction for women? You say that "in many developing nations, women are unfairly rejected from credit and financial institutions, due to being perceived as high-risk by lenders". Can you say how this might affect the loan prediction task? Can you add links to citations for this (and other claims throughout)? Is this use case intended to be situated in a low- or middle-income country? Particularly one that has a history of unfairly rejecting women from credit card applications? 

The "who will be harmed" section would be a good place to add more detail, but even from the beginning of the notebook it might be helpful for readers to have just a few words as to why you're evaluating for gender-based disparities. You have a line about racial disparities in lending, which I suppose begs the question why we aren't doing this analysis for that feature (other than that it's not in the dataset)? 

In that section, the numbered list starts with (1) "who will be harmed" and (2) "identify the types of harm", but then in the cell itself, it starts with "understanding what types of harms we anticipate" and focuses on allocation harms, before going on to discuss different demographic groups who might experience those harms. I know these are highly related steps, but it would be good to be consistent in the sequencing to not confuse readers.

**More clarity on the problem formulation would be helpful.** 
Specifically, making sure that we're consistent in referring to false negatives and false positives. The positive class here is referring to a default (i.e., failure to pay), so in line 619 (auxiliary harm #2) and line 635 (long-term effects), the false positive and false negative should be swapped. Note that this might occur in other places, these were just the two I caught, so you should double-check for consistency throughout. 

Perhaps relatedly, I didn't understand the renamed column names for PAY_0 and PAY_1. Aren't "default payment next month" and "default" the same thing? Should one of these be something like "bill paid next month" and "default next month". I suppose somewhere early on we should define "default" since that may not be a term all readers outside of finance will know already. 

I may be confused, but there's a lot of references to "loans" but, if I understand correctly, the dataset is of people paying off monthly credit card bills (or defaulting on them). Are these the same thing? Is the "loan" here referring to the credit card bills? I'm not sure that's a common usage, so if this is the technical financial industry terminology, we should define that explicitly for people. As it is, it reads like it's talking about two separate things - one about credit card payments and the other about applications for loans (e.g., for houses, cars, tuition, etc).

**Some more minor questions** 
- In line 586, what is "depending on where they live" referring to? 
- What does the term "limit" for the synthetic feature refer to? Is this a commonly used term for this that readers will be familiar with? (and is this a common way of defining it? e.g., using a Gaussian distribution with those criteria) 
- In line 656, to what extent are we "defining the fairness metrics based on harms"? 
- Do we have any guidance on how they might determine an appropriate value of epsilon? 
- "Disparity is reduced, but overall performance of model is reduced. What is the correct trade-off?" What is meant by "correct" here? 
- "In [Chen et. al (2018), the authors analyze the *BISG* framework and demonstrate the method systemically over-estimates the true level of disparate impact." This seems like it needs another sentence or two explaining the implications of this. That is, what should readers do with this information?

**Typos** 
- "Ernest" should be "Ernst" 
- "Maritial" should be "marital" 
- "Before we start training a classifier model, we want to explore the dataset for any data distribution imbalances that may " (occur) 
- "We see most (??) around 60% of…" 
- In our lending context, both *false_negative_rate_disparities* and *false_positive_rate_disparities* result in fairness-related harms. Therefore, we aim to " (??) 
- "females", "feamles" should be "women" 
- "males" should be "men" 
- "When passing in a fairness *_constaint_* as a `Moment` 
- "is sensitive to our _chose_ `epsilon` value" 

And some others throughout, so it would be good to take a pass for typos. 
  
Thanks again!

LeJit commented:
Hey @mmadaio,

I finally got around to updating the notebook to reflect your feedback. A summary of the changes is below:

> It would be good to add more explanation as to why you're using "sex" as the sensitive feature.

I added a section at the beginning about the Apple credit card story where the model allocated lower credit limits to women compared to men, even within the same household.

> Perhaps relatedly, I didn't understand the renamed column names for PAY_0 and PAY_1. Aren't "default payment next month" and "default" the same thing?

Added a small explanation in the notebook about the re-naming. `PAY_0` is renamed to `PAY_1` to keep the naming consistent with the other features, and `default payment next month` is re-named to `default` to reduce verbosity of the feature name (I didn't want to deal with a feature that had spaces in its name).

> I may be confused, but there's a lot of references to "loans" but, if I understand correctly, the dataset is of people paying off monthly credit card bills (or defaulting on them).

I changed the "story" of the case study to be predicting loan default rates using the monthly credit card payment dataset, where we use failure to make monthly card payments as a sign of poor "creditworthiness". Hopefully, this addresses the discrepancy between the two tasks.

> What does the term "limit" for the synthetic feature refer to? Is this a commonly used term for this that readers will be familiar with? (and is this a common way of defining it? e.g., using a Gaussian distribution with those criteria)

I changed the synthetic feature to represent `Interest` rates. The story behind the feature is now that people who have defaulted on credit card payments are given higher interest rates, and there is more _variance_ in the interest rates allocated to women compared to men.

mmadaio reviewed (commented):
This is much clearer than the previous iteration, Manojit! Thanks for your hard work on this!

You define default as "default payment next month" in the feature table, but you may want to explicitly stat it in the prose description too, instead of just saying that defaulting is failing to pay it back "in time" (or just failing to pay it back period).

You may want to explain why 60% is not considered imbalanced, but 78% is. Are there certain heuristics you're using to make that decision?

What do you mean by "less historical uncertainty about credit lending to men"?

Why choose a Gaussian distribution for the synthetic `Interest` feature?

You may want to add a line explaining how the lower variance in interest rates for men translates to the different parameters for men/women in your Gaussian distribution, for readers who may not be aware. 

What is the `LIMIT` feature? Is that the old name for what is now the `INTEREST` feature?

The "Who will be harmed?" section appears to be mostly about "Types of harm experienced". I would move the relevant parts of that section into Types of harm, and some of it seems like it may be redundant and need to be deleted. Do you have anything more to say about who might be harmed? Also, the order of the sections in the numbered list (Types of harms; who will be harmed) is different from that in the following sections. 

Will all readers know what a credit score is and why (and how) people's credit scores would be affected by loan rejections or defaults? I'm not sure that's necessarily true for readers outside of the US. You may want to link to an explainer and add a few words explaining it. (you have some of this in negative impact of [do you mean "_on_"?] credit score, but not when you first introduce credit scores in the prior section)

Why not develop a weighting mechanism for false negatives and false positives here, if that is recommended "in practice"?

Will your readers know what the ThresholdOptimizer does? You talk about a key limitation of it, but it would be good to add a sentence (or less) explaining it. In the Reducations section, you describe what the ThresholdOptimizer does ("tranform the model's decision boundary to satisfy our fairness constraints"), but this should be in the section on the ThresholdOptimizer.

In the last paragraph before the conclusion, you describe the impact that the post-processing and reductions approaches have on the model's performance disparities. However, you may want to provide a brief intuitive summary of the implications of what this means in terms of impacts on loan applicants. This will help readers contextualize the Conclusion section - and why it matters that they've reduced equalized odds without a drastic increase in balanced error score. 

I'm not sure that the "Other considerations" fit here at the end. Are there other places (e.g., earlier in the doc) that they might fit better? Or maybe mention them earlier and link to this section?

This is probably on your list, but you may want to take a pass for typos.

adrinjalali reviewed (changes requested):
IIRC we decided not to add new notebooks to the repo, and the old ones should also be removed. 

Also, now we have pre-commit hooks and you can enable them to fix your styling.

adrinjalali commented:
> I agree that we should discourage creating new notebooks in .ipynb, but I'm viewing this specific PR as a replacement of an existing notebook, so I'm okay with it being .ipynb. This is being linked to from our landing page, so I'm in favor of expediency here.

Then I'd like to remove the old one in this PR and add a new `.py` file instead.

MiroDudik commented on notebooks/Credit Loan Example.ipynb:
```diff
@@ -0,0 +1,1861 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this case study, we walked through the process of assessing a credit decision model for gender-related performance disparities. Our analysis follows closely from the work done in [*Dudik et al.* (2020)](https://www.microsoft.com/en-us/research/uploads/prod/2020/09/Fairlearn-EY_WhitePaper-2020-09-22.pdf) where we used the *Fairlearn* toolkit to perform an audit of a fairness-unaware tree-b ...
```
```suggestion
    "In this case study, we walked through the process of assessing a credit decision model for gender-related performance disparities. Our analysis follows closely the work done in the [*Microsoft/EY white paper*](https://www.microsoft.com/en-us/research/uploads/prod/2020/09/Fairlearn-EY_WhitePaper-2020-09-22.pdf) where they used the *Fairlearn* toolkit to perform an audit of a fairness-unaware tree-based model. We applied a *post-processing* and *reductions* mitigation techniques to mitigate the *equalized odds difference* in our model.\n",
```

MiroDudik commented on notebooks/Credit Loan Example.ipynb:
```diff
@@ -0,0 +1,1861 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Fairness Under Unawareness"
```
```suggestion
    "### Fairness Under Unawareness"
```

MiroDudik commented on notebooks/Credit Loan Example.ipynb:
```diff
@@ -0,0 +1,1861 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Conclusion"
```
```suggestion
    "## Conclusion and Discussion"
```

MiroDudik commented on notebooks/Credit Loan Example.ipynb:
```diff
@@ -0,0 +1,1861 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for index, row in performance_df_masked.iterrows():\n",
```
change the label to something like "eps=0.01, idx=10"

LeJit replied:
Done. Some of the labels overlap now, but I don't think it's a big deal. Most of them are still readable.

MiroDudik commented:
> > I agree that we should discourage creating new notebooks in .ipynb, but I'm viewing this specific PR as a replacement of an existing notebook, so I'm okay with it being .ipynb. This is being linked to from our landing page, so I'm in favor of expediency here.
> 
> Then I'd like to remove the old one in this PR and add a new `.py` file instead.

I've opened #1108 to discuss this point...

LeJit commented on notebooks/Credit Loan Example.ipynb:
```diff
@@ -0,0 +1,1861 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Binary Credit Loan Decisions with Unfairness Mitigation"
```
Changed.

LeJit commented on notebooks/Credit Loan Example.ipynb:
```diff
@@ -0,0 +1,1861 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "A key facet of Responsible Machine Learning is responsible documentation practices. *Mitchell et al. (2019)* proposed the model card framework for documentating and reporting model training details and deployment considerations. \n",
```
Done.

LeJit commented on notebooks/Credit Loan Example.ipynb:
```diff
@@ -0,0 +1,1861 @@
+   "source": [
+    "A key facet of Responsible Machine Learning is responsible documentation practices. *Mitchell et al. (2019)* proposed the model card framework for documentating and reporting model training details and deployment considerations. \n",
+    "\n",
+    "A *model card* contains sections for documenting training and evaluation dataset descriptions, ethical concerns, and quantitative evaluation summaries. In practice, we would ideally create a model card for our model before deploying it in production. Although we will not be producing a model card in this case study, interested readers can learn more about creating model cards using the *Mode ...
```
Done.

LeJit commented on notebooks/Credit Loan Example.ipynb:
```diff
@@ -0,0 +1,1861 @@
+    "- `EDUCATION`: Highest level of education achieved by the applicant.\n",
+    "- `MARRIAGE`: Marital status of the applicant.\n",
+    "\n",
+    "We will encode these variables as categorical features."
```
Done.

LeJit commented:
@MiroDudik I made all the changes you requested. You can take another look to see if there is anything else I need to change. I also deleted the old notebook. 

I'll add the new unit tests in a later commit.

adrinjalali commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1001 @@
+# Fairness and credit lending in the US
+# -----------------------------------------
+
+# %% 
+# In 2019, Apple received backlash on social media after its newly launched _Apple Card_ product appeared to `offer higher credit limits to men compared to women <https://edition.cnn.com/2019/11/12/business/apple-card-gender-bias/index.html>`_. In multiple cases, married couples found the husband received a credit limit that was 10-20x higher than the wife's even when the couple had joint assets.
+
+# %% 
+# From a regulatory perspective, financial institutions that operate within the United States are subject to _legal regulations_ prohibiting discrimination on the `basis of race, gender, or other protected classes <https://www.govinfo.gov/content/pkg/USCODE-2011-title15/html/USCODE-2011-title15-chap41-subchapIV.htm>`_. With the increasing prevalence of automated decision-systems in the financial  ...
+
+# %% 
+# Although the two concepts are intertwined, algorithmic fairness is not the same concept as anti-discrimination law. An AI system can comply with anti-discrimination law while exhibiting fairness-related concerns. On the other hand, some fairness interventions may be illegal under anti-discrimination law. `Xiang and Raji  (2019) <https://arxiv.org/abs/1912.00761>`_` discuss the compatibilities a ...
+
+# %% 
+# *EY* case study
+# ==================
+
+# %% 
+# In this case study, we aim to replicate the work done in a `white paper <https://www.microsoft.com/en-us/research/uploads/prod/2020/09/Fairlearn-EY_WhitePaper-2020-09-22.pdf>`_, co-authored by _Microsoft_ and _EY_, on mitigating gender-related performance disparities in financial lending decisions. In their analysis, _Microsoft_ and _EY_ demonstrated how _Fairlearn_ could be used to measure and ...
+
+# %% 
+# Using a dataset of credit loan outcomes (whether an individual defaulted on a credit loan), we train a fairness-unaware model to predict the likelihood an individual will default on a given loan. We use the *Fairlearn* toolkit for assessing the fairness of our model, according to several metrics. Finally, we perform two unfairness mitigation strategies on our model and compare the results to ou ...
+
+# %% 
+# Because the dataset used in the white paper is not publicly available, we will introduce a semi-synthetic feature into an existing publicly available dataset to replicate the outcome disparity found in the original dataset.
```
As long as there are no code, these can be in a single cell, we don't need to start a new cell for each paragraph with `# %%`. It'd also be nice to wrap the long lines to shorter ones, and you can use `alt+q` in emacs, or the same key in VSCode if you install the *Rewrap* extension to wrap the lines.

adrinjalali commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1001 @@
+
+# %%
+def plot_group_metrics_with_error_bars(metricframe, metric, error_name):
+  """
```
we should probably use 4 spaces indentation everywhere.

adrinjalali reviewed (commented):
You can probably fix a lot of formatting issues by running black on the code (or by enabling pre-commit hooks on your local clone).

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+    - names: The names of the selected metrics
+
+    Returns:
+        Metricframe: The concatenation of the two MetricFrame, resricted to the metrics specified.
```
```suggestion
        Metricframe: The concatenation of the two MetricFrames, restricted to the metrics specified.
```

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+    Concatenates the results of two MetricFrames along a subset of metrics.
+
+    Parameters:
+    - mframe_1: First metricframe for comparison
+    - mframe_2: Second metricframe for comparison
```
```suggestion
    - mframe_1: First MetricFrame for comparison
    - mframe_2: Second MetricFrame for comparison
```

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+    estimator, epsilon, X_train, y_train, A_train
+):
+    """
+    Instantiate and trains an ExponentiatedGradient model on the
```
```suggestion
    Instantiates and trains an ExponentiatedGradient model on the
```

romanlutz replied:
In fact, perhaps :code:`ExponentiatedGradient` or even :class:`ExponentiatedGradient` are probably better.

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+        - Epsilon: Float representing maximum difference bound for the fairness Moment constraint
+
+    Returns:
+        - Predictors: List of inner model predictors learned by the ExponentiatedModel
```
What's `ExponentiatedModel`?

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+# - 3.) Choose the model with smallest `equalized_odds` difference.
+
+# %%
+#  Within the context of fair lending in the United States, if a financial
```
```suggestion
# Within the context of fair lending in the United States, if a financial
```

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+# %%
+# We see both of our fairness-aware models yield a slight decrease in the
+# *balanced_accuracy* for *male applicants* compared to our fairness-unaware
+# model. In the *reductions* model, we see a decrease in the *false positive
```
I don't think the `*` can wrap around like this.

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+# deploying it in production. Although we will not be producing a model card in
+# this case study, interested readers can learn more about creating model cards
+# using the *Model Card Toolkit* from the `Fairlearn PyCon 2022 tutorial
+# <https://colab.research.google.com/github/LeJit/talks/blob/pycon_2022/2022_pycon/pycon-2022-students.ipynb>`_.
```
Can we reference the fairlearn talks repo instead?

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+# <https://colab.research.google.com/github/LeJit/talks/blob/pycon_2022/2022_pycon/pycon-2022-students.ipynb>`_.
+
+# %%
+# Fairness Under Unawareness
```
```suggestion
# Fairness under unawareness
```

LeJit replied:
_Fairness Under Unawareness_ (with that capitalization structure) is the literal name of the paper I'm referencing in this section .

romanlutz commented on fairlearn/datasets/_fetch_credit_card.py:
```diff
@@ -0,0 +1,76 @@
+
+def fetch_credit_card(*, cache=True, data_home=None,
+                      as_frame=False, return_X_y=False):
+    """Load the 'Default of Credit Card clients" dataset (binary classification).
```
either single or double quotes, not both 🙂

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+# card default outcomes for 30,000 applicants with information pertaining to an
+# applicant's payment history and bill statements over a six-month period from
+# April 2005 to September 2005, as well as demographic information, such as
+# *sex*, *age*, *marital status*, and *education level* of the applicant. A full
```
The question that always comes to mind for me in such situations is "What about gender?" and "What made you choose sex instead of gender?"
There's an issue we filed #792 to document best practices when working with gender. If that existed yet I would have asked you to reference it, but alas... 
Perhaps a sentence saying that we didn't collect this data, but if we had we may have included other sensitive/demographic columns as well ? Wdyt?

romanlutz reviewed (commented):
This looks great. Mostly adding nitpicky comments. Have you considered #418 in this context? I wonder if there was anything mentioned there that would be useful to include here.

LeJit commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1359 @@
+
+
+# %%
+def false_positive_error(y_true, y_pred):
+    """
+    Computes the standard error for the false positive rate estimate.
+    """
+    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
+    return compute_error_metric(fp, tn + fp)
+
+
+def false_negative_error(y_true, y_pred):
+    """
+    Computes the standard error for the false negative rate estimate.
+    """
+    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
+    return compute_error_metric(fn, fn + tp)
+
+
+def balanced_accuracy_error(y_true, y_pred):
+    """
+    Computes the standard error for the balanced accuracy estimate.
+    """
+    fpr_error, fnr_error = false_positive_error(
+        y_true, y_pred
+    ), false_negative_error(y_true, y_pred)
+    return np.sqrt(fnr_error**2 + fpr_error**2) / 2
```
As far as I know, we don't have functions defined to compute the _standard error_ for each of our metrics, just the metric values themselves.

romanlutz commented on examples/credit_loan_decisions.py:
```diff
@@ -0,0 +1,1329 @@
+# In the previous section, we took a fairness-unaware model and used the
+# `ThresholdOptimizer` to transform the model's decision boundary to satisfy our
+# fairness constraints. One key limitation of the `ThresholdOptimizer` is
+3  # needing access to our *sensitive_feature* during prediction time.
```
```suggestion
# needing access to our *sensitive_feature* during prediction time.
```

romanlutz reviewed (approved):
I went through the RST adjustments, removed the associated notebook tests, and remove the ipynb file. Looks fine to me now! @adrinjalali can you take a look since you've "requested changes." @LeJit please also take a look to make sure my changes make sense. And @MiroDudik may want to have a last read-through as one of the authors of the whitepaper.

romanlutz commented on notebooks/Binary Classification with the UCI Credit-card Default Dataset.ipynb:
```diff
@@ -1,710 +0,0 @@
-  {
-   "cell_type": "markdown",
-   "source": [
-    "# Simplifying this analysis using model comparison plots\n",
```
This seems to have been added somewhat recently via https://github.com/fairlearn/fairlearn/pull/947/files#diff-c9fd48863eb70895ec467c279a87b578a78a7318597e651561e57a52ef7d062e
Do we want to preserve this in some way?

adrinjalali reviewed (approved):
Thank you @romanlutz

romanlutz commented:
> Thank you @romanlutz

Of course! For context, we discussed this in the community call last week. Since there were quite a few ipynb-to-py (and md to rst) peculiarities that needed addressing I volunteered to take care of it as I've done this a few times. I suppose I should have provided a full page screenshot to make this easier to review:
![screencapture-file-C-Users-Roman-git-fairlearn-docs-build-html-auto-examples-plot-credit-loan-decisions-html-2022-12-19-11_42_18](https://user-images.githubusercontent.com/10245648/208476171-07bd87f8-aa7e-4d4f-812a-836c619bfbe6.png)
![screencapture-file-C-Users-Roman-git-fairlearn-docs-build-html-auto-examples-plot-credit-loan-decisions-html-2022-12-19-11_42_18-2](https://user-images.githubusercontent.com/10245648/208476173-a9488d44-96ee-4ecd-a49b-5e86af9a3402.png)


... and I already found a couple of `$` signs that I hadn't converted yet. 😆

romanlutz commented:
@LeJit why is seaborn in the dependencies? I don't see it used.

LeJit commented:
> @LeJit why is seaborn in the dependencies? I don't see it used.

@romanlutz There were some things with visualizing and comparing error rates across different models at the end that Miro and I wanted to, but we eventually decided to cut. You can remove the `seaborn` dependency and I believe the `CalibrationClassifierCV` import as well.

MiroDudik commented on examples/plot_credit_loan_decisions.py:
```diff
@@ -0,0 +1,1291 @@
+
+estimator.fit(X_train, y_train)
+
+# %% We compute the *binary predictions* and the *prediction probabilities* for
```
```suggestion
# %%
# We compute the *binary predictions* and the *prediction probabilities* for
```

MiroDudik commented on examples/plot_credit_loan_decisions.py:
```diff
@@ -0,0 +1,1291 @@
+
+# %%
+# Here, we can see all the inner models learned for each value of
+# :param:`epsilon`. With the :code:`ExponentiatedGradient` model, we specify an
```
```suggestion
# :code:`epsilon`. With the :code:`ExponentiatedGradient` model, we specify an
```

MiroDudik commented on notebooks/Binary Classification with the UCI Credit-card Default Dataset.ipynb:
```diff
@@ -1,710 +0,0 @@
-{
```
Besides deleting this file, also update the pointer from the landing page to point to the new example.

romanlutz replied:
Done! Pointing to https://fairlearn.org/main/auto_examples/plot_credit_loan_decisions.html which will exist after this PR.

MiroDudik commented on examples/plot_credit_loan_decisions.py:
```diff
@@ -0,0 +1,1293 @@
+# is changed to :code:`default` to reduce verbosity.
+
+# %%
+data_url = "http://archive.ics.uci.edu/ml/machine-learning-databases/00350/default%20of%20credit%20card%20clients.xls"
```
Did we want to use the newly added `fetch_credit_card` instead of this?

romanlutz replied:
The problem with `fetch_credit_card` is that these are the column names:
` x1   x2   x3   x4    x5   x6   x7   x8   x9  x10  x11       x12       x13       x14      x15      x16      x17      x18      x19      x20     x21      x22     x23`
Should we fix this in the definition of `fetch_credit_card`?

MiroDudik replied:
uff. this makes me worried about `fetch_credit_card`. i am fine merging as is, but i'll open an issue to fix the feature names. we have the same problem with `fetch_bank_marketing`.

MiroDudik replied:
See #1165

MiroDudik commented on examples/plot_credit_loan_decisions.py:
```diff
@@ -0,0 +1,1293 @@
+
+# %%
+metricframe_unmitigated.by_group[metrics_to_report].plot.bar(
+    subplots=True, layout=[3, 1], figsize=[12, 18]
```
I think it might look better if we replace this by the layout [1, 3] (with the appropriate figsize)? But I guess it slightly depends on whether we expect people to consume it on desktop or mobile. If we change it here, we should also change it elsewhere in the example.

romanlutz replied:
This is what it renders as right now. That looks fine to me(?) The screenshot is from my desktop, not mobile.

![image](https://user-images.githubusercontent.com/10245648/209249459-30d0f99a-cae4-4329-863e-aa17187df5b5.png)

MiroDudik replied:
you also need to adjust figsize... i've implemented in my commit below, so this is done.

MiroDudik commented on examples/plot_credit_loan_decisions.py:
```diff
@@ -0,0 +1,1293 @@
+
+
+# %%
+def aggregate_predictor_performances(
```
We should replace the chunk of the code starting roughly here by a call to:
https://fairlearn.org/v0.8/api_reference/fairlearn.metrics.html#fairlearn.metrics.plot_model_comparison

LeJit replied:
We're currently using `aggregate_predictor_performances` and `model_performance_sweep` function to determine which inner models of the `ExponentiatedGradient` are Pareto-dominated. If we were to replace this chuck of code with a call to `plot_model_comparison`, would we show the performance of all the inner models or still just the non-dominated ones.

I think we could replace the lines from *1031*-*1054* with this new function, if we really wanted.

MiroDudik replied:
You are right that things are intertwined... I was hoping that we could somehow just use is_pareto_efficient and filter_dominated_rows, but not the rest. I see two options: (1) leave things as they are, (2) use the plot_model_comparison (on all of the models, without the restricting to pareto optimal ones)--was it a mess to look at all of them?

LeJit replied:
I believe there are 35 inner models total (18 for one value of `epsilon`, 17 for the other), so it would be a bit cluttered with all the inner models.

MiroDudik replied:
ok. let's leave as is for now then.

MiroDudik commented on fairlearn/datasets/_fetch_credit_card.py:
```diff
@@ -0,0 +1,77 @@
+# Copyright (c) Fairlearn Contributors.
```
This is incomplete. We still need to update `fairlearn/datasets/__init__.py` and `test/unit/datasets/test_datasets.py`.

romanlutz replied:
Done.

MiroDudik reviewed (changes requested):
This is almost ready to go. The main essential changes:
* update the link from the fairlearn landing page
* finish up work on `fetch_credit_card` (alternatively, I'd be also okay keeping that for a separate PR)
The other comments are just small suggestions.

romanlutz commented:
I'll try to get to it today. I have a bunch of tiny fixes locally as well. I suspect the linter will make me do a whole lot of fixes, too.

MiroDudik commented:
You may want to merge PR #1163 to make the linter work again.

romanlutz commented on examples/plot_credit_loan_decisions.py:
```diff
@@ -0,0 +1,1293 @@
+#
+# In the previous section, we trained multiple :code:`ExponentiatedGradient` models
+# at different :code:`epsilon` levels and collected all the inner models learned by
+# this process. However, some of these inner models are better performance in
```
This sentence doesn't quite make sense. I changed it from "are better performance" to "have worse performance"
Wdyt @MiroDudik  @LeJit ?

MiroDudik replied:
i have edited this, so it's now fixed.

romanlutz commented:
This is now up-to-date with latest main and I've fixed all flake8 issues.
The remaining open issues afaict:
- we're not using `fetch_credit_card` because the column names are off (see corresponding comment)
- we're not using `plot_model_comparison`

These seem like easily fixable problems in small follow-up PRs that don't affect the example in terms of validity. If either of you (@MiroDudik, @LeJit ) want to take a stab at them please feel free to push your updates directly. I've done the part that I had promised 😄

LeJit commented:
I think it would be better to make those changes in follow-up PRs. I really can't think of any way to identify the Pareto dominated inner models that doesn't require us to compute the `balanced_accuracy` and `equalized_odds` for each of the inner models, so we're always going to have to do that inner models sweep.

Also, if we want an example using `plot_model_comparison`, I personally think the SciPy/PyCon tutorial would be a better place to incorporate that method since we're not filtering dominated classifiers in that example. What do you think @MiroDudik ?

MiroDudik commented:
i tried to deal with the issue of crashing build of the website, but the issue is that the sphinx gallery simply takes too long to build on CircleCI...

iofall commented on examples/plot_credit_loan_decisions.py:
```diff
@@ -0,0 +1,1268 @@
+        model during the training process.
+
+    """
+    exp_grad_est = ExponentiatedGradient(
+        estimator=estimator.named_steps["classifier"],
+        constraints=EqualizedOdds(difference_bound=epsilon),
+    )
+    # Is this an issue - Re-runs
+    exp_grad_est.fit(X_train, y_train, sensitive_features=A_train)
+    predictors = exp_grad_est.predictors_
+    return predictors
```
I did try to look into the CircleCI issue. It runs properly till we encounter the function `get_expgrad_models_per_epsilon` call.

The iterations get processed very slowly, not sure why it happens so quickly on local. Maybe we could limit it somehow, or make this part hard-coded at least for generating sphinx gallery examples.

romanlutz replied:
There's been a little bit of a discussion about this on [Discord](https://discordapp.com/channels/840099830160031744/872866935610703952/1060658509861109930), too. I'm still wondering if we should stick with CircleCI at all, assuming this issue wouldn't happen on GitHub Actions. But migrating is also work, of course.

MiroDudik replied:
I've been fiddling with this on and off, but still no luck.

MiroDudik replied:
Now is now resolved. See:
https://github.com/fairlearn/fairlearn/pull/1170#issuecomment-1377421962
